addons/onboarding/tests/test_onboarding.py

```diff
@@ -0,0 +1,136 @@
+        self.assert_onboarding_is_not_done(self.onboarding_1)
+
+        # Completing it sets onboarding state to done again
+        onboarding_1_step_3.action_set_just_done()
+        self.assert_onboarding_is_done(self.onboarding_1)
+
+        # If a company is added, onboarding is 'done'
+        company_3 = self.env.company.create({
+            'currency_id': self.env.ref('base.EUR').id,
+            'name': 'Another Test Company',
+        })
+        self.assert_onboarding_is_done(self.onboarding_1.with_company(company_3))
+
+        # Adding new step resets onboarding state to 'not_done'
+        self.env['onboarding.onboarding.step'].create({
+            'title': 'Test Onboarding 1 - Step 4',
+            'onboarding_id': self.onboarding_1.id,
+            'panel_step_open_action_name': 'action_fake_open_onboarding_step',
+        })
+
+        # Closing the panel still allows to track if all steps are completed
+        self.onboarding_1.action_close()
+        self.assertTrue(self.onboarding_1.current_progress_id.is_onboarding_closed)
+        self.assert_onboarding_is_not_done(self.onboarding_1)
+
+    def test_onboarding_completion_per_company(self):
+        """Checks the behavior of onboarding and step states in multi-company setting:
+        the onboarding state has to track the completion of each of its steps, global and
+        per-company, to determine if whether it is completed.
+        """
+        # Completing onboarding as company_1
+        self.assertEqual(self.env.company, self.company_1)
+
+        # Updating onboarding to per-company
+        self.onboarding_1.is_per_company = True
+        # Required after progress reset
+        self.onboarding_1._search_or_create_progress()
+
+        self.onboarding_1_step_1.action_set_just_done()
+        self.assert_step_is_done(self.onboarding_1_step_1)
+
+        self.onboarding_1_step_2.action_set_just_done()
+        self.assert_onboarding_is_done(self.onboarding_1)
+
+        # Completing onboarding as existing company_2
+        self.activate_company(self.company_2)
+        # First access from company_2
+        self.onboarding_1._search_or_create_progress()
+
+        # Blank state for company 2
+        self.assert_step_is_not_done(self.onboarding_1_step_1)
+        self.assert_onboarding_is_not_done(self.onboarding_1)
+
+        # But no change for company 1
+        self.assert_step_is_done(self.onboarding_1_step_1.with_company(self.company_1))
+        self.assert_onboarding_is_done(self.onboarding_1.with_company(self.company_1))
+
+        self.onboarding_1_step_1.action_set_just_done()
+        self.assert_step_is_done(self.onboarding_1_step_1)
+        self.assert_onboarding_is_not_done(self.onboarding_1)
+        self.onboarding_1_step_2.with_company(self.company_2).action_set_just_done()
+        self.assert_step_is_done(self.onboarding_1_step_2)
+        self.assert_onboarding_is_done(self.onboarding_1)
+
+        # create company_3
```

Comment:
Out of curiosity, is adding a third company adding coverage to the tests ? Did not read in details, just asking (if we can shorten a bit, well, otherwise that's pas grave).